File: 200-number-of-islands/number-of-islands.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  int numIslands(vector<vector<char>> &grid) {
    int n = grid.size(), m = grid[0].size();
    int cnt = 0;
    function<void(int x, int y)> traverse = [&](int x, int y) {
      if (x < 0 || y < 0 || x >= n || y >= m)
        return;
      if (grid[x][y] == '1')
        grid[x][y] = '0';
      else
        return;
      vector<int> dx = {0, 0, 1, -1}, dy = {1, -1, 0, 0};
      for (int i = 0; i < 4; i++) {
        traverse(x + dx[i], y + dy[i]);
      }
    };
    for (int i = 0; i < n; i++)
      for (int j = 0; j < m; j++)
        if (grid[i][j] == '1')
          cnt += 1, traverse(i, j);
    return cnt;
  }
};

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        cnt = 0
        m, n = len(grid), len(grid[0])
        vis = [[0] * n for _ in range(m)]
        def dfs(x, y):
            if x < 0 or y < 0 or x >= m or y >= n:
                return
            if grid[x][y] == '0':
                return
            if vis[x][y]:
                return
            vis[x][y] = 1
            dfs(x + 1, y), dfs(x - 1, y)
            dfs(x, y + 1), dfs(x, y - 1)
        for i in range(m):
            for j in range(n):
                if vis[i][j]:
                    continue;
                if grid[i][j] == '1':
                    cnt = cnt + 1
                    dfs(i, j)
        return cnt